src/Compilers/Core/Portable/InternalUtilities/ConcurrentDictionaryExtensions.cs

```diff
@@ -19,8 +21,22 @@ internal static class ConcurrentDictionaryExtensions
         {
             if (!dict.TryAdd(key, value))
             {
-                throw new System.ArgumentException("adding a duplicate");
+                throw new ArgumentException("adding a duplicate", nameof(key));
             }
         }
+
+        public static TValue GetOrAdd<TKey, TArg, TValue>(this ConcurrentDictionary<TKey, TValue> dictionary, TKey key, Func<TKey, TArg, TValue> valueFactory, TArg factoryArgument)
+            where TKey : notnull
+        {
+#if NETCOREAPP
+            return dictionary.GetOrAdd(key, valueFactory, factoryArgument);
+#else
+            if (dictionary.TryGetValue(key, out var value))
+                return value;
+
+            using var _ = PooledDelegates.GetPooledFunction(valueFactory, factoryArgument, out var boundFunction);
```

Comment:
Where is the `boundFunction` actually used?

Reply:
I was afraid I might do this, hence adding a test for it 🤣

Reply:
Amended the first commit to correct this line.